cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -0,0 +1,270 @@
+    int argsStartIndex = getArgsStartIndex(str, macroStr);
+
+    // determine whether to use a macro function or a property lookup
+    if (argsStartIndex >= 0) {
+      return getMacroFunctionMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr, argsStartIndex);
+    } else {
+      return getMacroLookupMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr);
+    }
+  }
+
+  private static MacroMetadata getMacroFunctionMetadata(@Nullable MacroEvaluator macroEvaluator,
+                                                        @Nullable Set<String> macros, int startIndex, int endIndex,
+                                                        String macroStr, int argsStartIndex) {
+    // check for closing ")" and allow escaping "\)" TODO: and doubly-escaping "\\)"
+    int closingParenIndex = getClosingParenIndex(macroStr);
+    if (closingParenIndex < 0 || !macroStr.endsWith(")")) {
+      throw new InvalidMacroException(String.format("Could not find enclosing ')' for macro arguments in '%s'.",
+                                                    macroStr));
+    }
+
+    // arguments and macroFunction are expected to have escapes replaced when being evaluated
+    String arguments = macroStr.substring(argsStartIndex + 1, macroStr.length() - 1);
+    String macroFunction = macroStr.substring(0, argsStartIndex);
+    if (macroEvaluator != null) {
+      arguments = replaceEscapedSyntax(arguments);
+      macroFunction = replaceEscapedSyntax(macroFunction);
+    }
+    String[] args = COMMA.split(arguments);
+    if (macros != null) {
+      for (String argument : args) {
+        macros.add(argument);
+      }
+    }
+    return createMacroMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr, macroFunction, args);
+  }
+
+  private static MacroMetadata getMacroLookupMetadata(@Nullable MacroEvaluator macroEvaluator,
+                                                      @Nullable Set<String> macros, int startIndex, int endIndex,
+                                                      String macroStr) {
+    // property is expected to have escapes replaced only when being evaluated
+    if (macroEvaluator != null) {
+      macroStr = replaceEscapedSyntax(macroStr);
+    }
+    if (macros != null) {
+      macros.add(macroStr);
+    }
+    return createMacroMetadata(macroEvaluator, macros, startIndex, endIndex, macroStr, null, null);
+  }
+
+  /**
+   * Creates MacroMetadata object based on path substitute was called from (containsMacro, getMacros, or substitute)
+   */
+  private static MacroMetadata createMacroMetadata(@Nullable MacroEvaluator macroEvaluator,
+                                                   @Nullable Set<String> macros, int startIndex, int endIndex,
+                                                   String macroStr, @Nullable String macroFunction,
+                                                   @Nullable String[] args) {
+    if (macroEvaluator != null) {
```

Comment:
all of these null checks for different code paths make it difficult to understand whats going on. Rather than having getMacros, containsMacros, substituteMacros, and different null checks handling each scenario, I think it would be simpler to understand if you have a single method `ParsedProperty parse(String str)`.

ParsedProperty would contain the substituted string, a list of lookups that were done, and a list of functions that were done. You could pass in a MacroEvaluator in the constructor. If you are only interested in what macros are in the string, you could pass in a DummyEvaluator that always returns an empty string.